roles/testnode/files/libexec/smart.sh

```diff
@@ -0,0 +1,254 @@
+### Variables ###
+numdrives=0 # total number of drives (or RAID slots) discovered
+failingdrives=0 # Number of failed, failing, and/or missing drives
+rc=0 # Return code for nagios
+unknownmsg="Unknown error" # Fallback message for UNKNOWN return code output
+
+# Array of messages indicating drive health.  Output after nagios status.
+declare -a messages
+
+### Functions ###
+
+# Gather smart data for drives behind Areca RAID controller
+# NOTE: Only works when individual disks set to JBOD for now
+areca_smart ()
+{
+  numdrives=$(sudo cli64 disk info | grep -E "Slot#[[:digit:]]" | wc -l)
+  # Loop through all disks not marked as 'N.A.' or 'Failed'
+  for slot in $(sudo cli64 disk info | grep -E "Slot#[[:digit:]]" \
+  | grep -v 'N.A.\|Failed' | grep -o "Slot#[[:digit:]]" | cut -c6-)
+  do
+    failed=false
+    output=$(sudo cli64 disk smart drv=$slot \
+    | grep -E "^  "5"|^"197"|^"198"" | awk '{ print $(NF-1) }')
+    outputcount=$(echo $output | wc -w)
+    # Determine drive letter given slot number
+    dl=$(areca_bay_to_letter $slot)
+    # Only continue if we received 3 SMART data points
+    if [ "$outputcount" = "3" ]
+    then
+      read reallocated pending uncorrect <<< $output
+      if [ "$reallocated" != "0" ]
+      then
+        messages+=("Drive $slot $dl has $reallocated reallocated sectors")
+        failed=true
+        # A small number of reallocated sectors is OK
+        if [ "$reallocated" -le 5 ]
+        then
+          rc=1 # Warn if <= 5
+        else
+          rc=2 # Crit if >5
+        fi
+      fi
+      if [ "$pending" != "0" ]
+      then
+        messages+=("Drive $slot $dl has $pending pending sectors")
+        failed=true
+        rc=2
+      fi
+      if [ "$uncorrect" != "0" ]
+      then
+        messages+=("Drive $slot $dl has $uncorrect uncorrect sectors")
+        failed=true
+        rc=2
+      fi
+    elif [ "$outputcount" != "3" ]
+    then
+      messages+=("Drive $slot returned $outputcount of 3 expected attributes")
+      unknownmsg="SMART data could not be read for one or more drives"
+      rc=3
+    fi
+    # Make sure drives with multiple types of bad sectors only get counted once
+    if [ "$failed" = true ]
+    then
+      let "failingdrives+=1"
+    fi
+  done
+}
+
+# Correlate Areca drive bay to drive letter
+areca_bay_to_letter ()
+{
+  # Get S/N according to RAID controller given argument $1 (slot #)
+  areca_serial=$(sudo cli64 disk info drv=$1 | grep 'Serial Number' \
+  | awk '{print $NF}')
+  # Loop through and get S/N according to smartctl given drive name
+  for dl in $(cat /proc/partitions | grep -w 'sd[a-z]\|sd[a-z]\{2\}' \
+  | awk '{ print $NF }')
+  do
+    smart_serial=$(sudo smartctl -a /dev/$dl | grep "Serial number" \
+    | awk '{ print $NF }')
+    # If cli64 and smartctl find a S/N match, return drive letter
+    if [ "$areca_serial" = "$smart_serial" ]
+    then
+      echo "($dl)"
+    fi
+  done
+}
+
+# Standard SATA/SAS drive smartctl check
+normal_smart ()
```

Comment:
I thought about this and you're right.

The main reason I've kept them separate so far is in case we have to go back to the legacy way of polling SMART data from disks behind an Areca controller.  (`sudo smartctl -a -d areca,$drive $scsidev`)

I also need to tweak `areca_smart ()` a bit more so it can handle disks in part of a hardware RAID.